### PR TITLE
[WIP] Cuckoo Cycle PoW as separate OS process, and configurable

### PR DESCRIPTION
**Pull request by lucafavatella**
Supersedes PR #350.

Finishes:
* [PT-152519020](https://www.pivotaltracker.com/n/projects/2124891/stories/152519020) - miner death handling
* [PT-152519057](https://www.pivotaltracker.com/n/projects/2124891/stories/152519057) - configurable 16 vs. 30
* [PT-152519288](https://www.pivotaltracker.com/n/projects/2124891/stories/152519288) - configurable mean vs. lean

It does not finish:
* [PT-152518745](https://www.pivotaltracker.com/n/projects/2124891/stories/152518745) - miner as separate OS process
  * ... because of needed refinement of PoW nonce size in code and/or API and/or wiki. @akorosmezey to confirm missing steps if any.

Known issues:
* CI job of release of testing script on OSX (newly introduced job, under review) fails as no block gets mined (see [sample log from old build](https://travis-ci.org/aeternity/epoch/jobs/300003870#L1646)). Issue may be in the fragile sed command execution at the beginning of the script. It passes (quickly) on Ubuntu;
  * @akorosmezey may you please ensure CI passes?
* [PT-152796665](https://www.pivotaltracker.com/story/show/152796665)

### CODE REVIEW

File: apps/ae_cuckoo_cycle_pow_executables/Makefile [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+REPO = https://github.com/aeternity/cuckoo.git
+COMMIT = fddbf60
+
+.PHONY: all
+all: \
+	priv/lib/libblake2b.so \
+	$(EXECUTABLES)
+
+.PHONY: clean
+clean:
```
Comment (happi) [inline apps/ae_cuckoo_cycle_pow_executables/Makefile]: What files are there in c_src? It looks strange that clean would delete source code. I would think dist clean would do this bu clean would only delete generated files?
Are there only generated files in c_src?

File: apps/aecore/src/aec_pow_cuckoo.erl [hunk 2]
```diff
@@ -95,44 +92,251 @@ verify(Data, Nonce, Evd, Target) when is_list(Evd) ->
...
+            {error, no_value};
+        {'DOWN', OsPid, process, _, Reason} ->
+            %% Abnormal termination
+            Reason2 = case Reason of
+                          {exit_status, ExStat} -> exec:status(ExStat);
+                          _                     -> Reason
+                      end,
+            ?error("OS process died: ~p~n", [Reason2]),
+            {error, {execution_failed, Reason2}};
+        Other ->
```
Comment (happi) [inline apps/aecore/src/aec_pow_cuckoo.erl]: Unused Other (in non debug case) replace by _Other.

File: apps/ae_cuckoo_cycle_pow_executables/test/ae_cc_pow_executables_SUITE.erl [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    [smoke_test].
+
+smoke_test(_Config) ->
+    LibDir = ?TEST_MODULE:lib_dir(),
+    MinBin = ?TEST_MODULE:bin("lean16"),
+    VerBin = ?TEST_MODULE:bin("verify16"),
+    LdLibPathVarName = ld_lib_path_var_name(),
+    Cmd =
+        io_lib:format(
+          "env ~s='~s' '~s' -n 66 | grep '^Solution ' | env ~s='~s' '~s' -n 66 | grep '^Verified '",
```
Comment (uwiger) [inline apps/ae_cuckoo_cycle_pow_executables/test/ae_cc_pow_executables_SUITE.erl]: I would prefer using the ability to break the string across multiple lines, e.g.:

```erlang
io_lib:format("env ~s='~s' '~s' -n 66 "
              "| grep '^Solution '"
              "| env ~s='~s' '~s' -n 66 "
              "| grep '^Verified '",
              [...])
```

**Comment by happi** [comment]
Please fix:

apps/aecore/src/aec_pow_cuckoo.erl:280: Warning: variable 'Msg' is unused
apps/aecore/src/aec_pow_cuckoo.erl:301: Warning: variable 'Msg' is unused
apps/aecore/src/aec_pow_cuckoo.erl:313: Warning: variable 'Reason' is unused
apps/aecore/src/aec_pow_cuckoo.erl:316: Warning: variable 'R' is unused

File: apps/ae_cuckoo_cycle_pow_executables/test/ae_cc_pow_executables_SUITE.erl (hunk 3 above)
Comment (akorosmezey) [reply apps/ae_cuckoo_cycle_pow_executables/test/ae_cc_pow_executables_SUITE.erl]: Addressed in next commit 6c57766

**Comment by akorosmezey** [comment]
@happi unused vars addressed in the next commit 6c57766

**Comment by sennui** [comment]
it looks like it is not rebased. I see Ulf's changes.

File: apps/ae_cuckoo_cycle_pow_executables/src/ae_cuckoo_cycle_pow_executables.app.src [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{application, ae_cuckoo_cycle_pow_executables,
```
Comment (sennui) [inline apps/ae_cuckoo_cycle_pow_executables/src/ae_cuckoo_cycle_pow_executables.app.src]: this is way too long name. Also please follow the convention without underscore aecuckoo is fine.

**Comment by lucafavatella** [comment]
Closing. Opened new one (#363) rebased on master and addressing the following issues in this PR:
* mean16 miner does not work on OSX Travis CI machines. Move tests to lean16 miner;
* Name ae_cuckoo_cycle_pow_executables too long -> aecuckoo;
* Sped up CI.

**Pull request closed without merging**